Mathlib/LinearAlgebra/Matrix/FixedDetMatrices.lean

```diff
@@ -52,4 +53,210 @@ lemma smul_coe (m : R) (g : SpecialLinearGroup n R) (A : FixedDetMatrix n R m) :
+
+lemma reps_zero_empty : (reps 0) = ∅ := by
+  rw [reps]
+  refine Set.eq_empty_iff_forall_not_mem.mpr fun A h ↦ ?_
+  obtain ⟨h₁, h₂, -, h₄⟩ := Set.mem_setOf.mpr h
+  simp only [eq_zero_of_ne_zero_of_mul_left_eq_zero h₂.ne' <| A_c_eq_zero h₁, abs_zero] at h₄
+  exact ((abs_nonneg _).trans_lt h₄).false
+
+noncomputable instance reps_fintype (k : ℤ) : Fintype (reps k) := by
+  by_cases hk : k = 0
+  · rw [hk, reps_zero_empty]
+    exact Set.fintypeEmpty
+  · let H := Finset.Icc (-|k|) |k|
+    let H4 := Fin 2 → Fin 2 → H
+    apply Fintype.ofInjective (β := H4)
+      (f := fun (M : reps k) ↦ fun i j ↦ ⟨M.1.1 i j, reps_entries_le_m' hk M.2 i j⟩)
+    intro M N h
+    ext i j
+    simpa only [Subtype.mk.injEq] using congrFun₂ h i j
+
+lemma reduce_mem_reps {m : ℤ} (hm : m ≠ 0) : ∀ A : Δ m, reduce A ∈ reps m := by
+  apply reduce_rec
+  · intro A h
+    by_cases h1 : 0 < A.1 0 0
+    · simp only [reps, reduce_of_pos h h1, zpow_neg, Set.mem_setOf_eq, smul_coe,
+        coe_inv, coe_T_zpow, adjugate_fin_two_of, neg_zero, cons_mul, Nat.succ_eq_add_one,
+        Nat.reduceAdd, empty_mul, Equiv.symm_apply_apply, of_apply, cons_val', vecMul,
+        cons_dotProduct, vecHead, one_mul, vecTail, Function.comp_apply, Fin.succ_zero_eq_one,
+        neg_mul, dotProduct_empty, add_zero, zero_mul, zero_add, empty_val', cons_val_fin_one,
+        cons_val_one, cons_val_zero, lt_add_neg_iff_add_lt, le_add_neg_iff_add_le]
+      refine ⟨abs_eq_zero.mp h, ?_, ?_, ?_⟩
+      · simp only [abs_eq_zero.mp h, mul_zero, h1]
+      · exact Int.ediv_mul_le _ <| A_d_ne_zero h hm
+      · rw [mul_comm, ← Int.sub_eq_add_neg, ← Int.emod_def]
+        apply le_trans _ (Int.emod_lt (A.1 0 1) (A_d_ne_zero h hm))
+        rw [abs_eq_self.mpr (Int.emod_nonneg (A.1 0 1) (A_d_ne_zero h hm))]
+    · simp only [reps, reduce_of_not_pos h h1, Int.ediv_neg, neg_neg, smul_def, ←
+        mul_assoc, S_mul_S_eq, neg_mul, one_mul, coe_T_zpow, mul_neg, cons_mul, Nat.succ_eq_add_one,
+        Nat.reduceAdd, empty_mul, Equiv.symm_apply_apply, neg_of, neg_cons, neg_empty,
+        Set.mem_setOf_eq, of_apply, cons_val', Pi.neg_apply, vecMul, cons_dotProduct, vecHead,
+        vecTail, Function.comp_apply, Fin.succ_zero_eq_one, dotProduct_empty, add_zero, neg_add_rev,
+        zero_mul, zero_add, empty_val', cons_val_fin_one, cons_val_one, neg_eq_zero, cons_val_zero,
+        lt_add_neg_iff_add_lt, le_add_neg_iff_add_le, abs_neg]
+      refine ⟨abs_eq_zero.mp h, ?_, ?_, ?_⟩
+      · simp only [abs_eq_zero.mp h, mul_zero, neg_zero, Int.lt_iff_le_and_ne, ne_eq]
+        refine ⟨not_lt.mp h1, A_a_ne_zero (abs_eq_zero.mp h) hm⟩
+      · rw [le_neg]
+        exact Int.ediv_mul_le _ <| A_d_ne_zero h hm
+      · rw [mul_comm, add_comm, ← Int.sub_eq_add_neg, ← Int.emod_def]
+        apply le_trans _ (Int.emod_lt (-A.1 0 1) (A_d_ne_zero h hm))
+        rw [abs_eq_self.mpr (Int.emod_nonneg (-A.1 0 1) (A_d_ne_zero h hm))]
```

Comment:
```suggestion
    have hd := A_d_ne_zero h hm
    by_cases h1 : 0 < A.1 0 0
    focus simp only [reduce_of_pos h h1]
    swap
    focus simp only [reduce_of_not_pos h h1]
    swap
    all_goals
      simp only [reps, zpow_neg, Set.mem_setOf_eq, smul_coe, coe_inv, coe_T_zpow,
        adjugate_fin_two_of, neg_zero, cons_mul, Nat.succ_eq_add_one, Nat.reduceAdd, empty_mul,
        Equiv.symm_apply_apply, of_apply, cons_val', vecMul, cons_dotProduct, vecHead, one_mul,
        vecTail, Function.comp_apply, Fin.succ_zero_eq_one, neg_mul, dotProduct_empty, add_zero,
        zero_mul, zero_add, empty_val', cons_val_fin_one, cons_val_one, cons_val_zero,
        lt_add_neg_iff_add_lt, le_add_neg_iff_add_le, Int.ediv_neg, neg_neg, smul_def, ← mul_assoc,
        S_mul_S_eq, mul_neg, neg_of, neg_cons, neg_empty, Pi.neg_apply, neg_add_rev, neg_eq_zero,
        abs_neg, ← le_neg, abs_eq_zero.mp h, true_and, mul_zero, lt_neg]
    · refine ⟨h1, Int.ediv_mul_le _ hd, ?_⟩
      rw [mul_comm, ← Int.sub_eq_add_neg, ← Int.emod_def, abs_eq_self.mpr <| Int.emod_nonneg _ hd]
      exact Int.emod_lt _ hd
    · refine ⟨?_, Int.ediv_mul_le _ hd, ?_⟩
      · simp only [Int.lt_iff_le_and_ne]
        exact ⟨not_lt.mp h1, A_a_ne_zero (abs_eq_zero.mp h) hm⟩
      · rw [mul_comm, add_comm, ← Int.sub_eq_add_neg, ← Int.emod_def,
         abs_eq_self.mpr <| Int.emod_nonneg _ hd]
        exact Int.emod_lt _ hd
```
Trying to write this a bit more compactly (and making use of the fact that both branches are quite similar).